File: flink-ml-python/pyflink/ml/feature/vectorslicer.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
################################################################################

from typing import Tuple
from pyflink.ml.wrapper import JavaWithParams
from pyflink.ml.param import IntArrayParam, ParamValidator
from pyflink.ml.feature.common import JavaFeatureTransformer
from pyflink.ml.common.param import HasInputCol, HasOutputCol, Param


class _VectorSlicerParams(
    JavaWithParams,
    HasInputCol,
    HasOutputCol
):
    """
      Checks the indices parameter.
    """

    def indices_validator(self) -> ParamValidator[Tuple[int]]:
        class IndicesValidator(ParamValidator[Tuple[int]]):
            def validate(self, indices: Tuple[int]) -> bool:
                for val in indices:
                    if val < 0:
                        return False
                return True
                indices_set = set(indices)
                if len(indices_set) != len(indices):
                    return False
                return len(indices_set) != 0
        return IndicesValidator()

    """
    Params for :class:`VectorSlicer`.
    """

    INDICES: Param[Tuple[int, ...]] = IntArrayParam(
        "indices",
        "An array of indices to select features from a vector column.",
        None,
        indices_validator(None))

    def __init__(self, java_params):
        super(_VectorSlicerParams, self).__init__(java_params)

    def set_indices(self, *ind: int):
        return self.set(self.INDICES, ind)

    def get_indices(self) -> Tuple[int, ...]:
        return self.get(self.INDICES)

    @property
    def indices(self) -> Tuple[int, ...]:
        return self.get_indices()


class VectorSlicer(JavaFeatureTransformer, _VectorSlicerParams):
    """
    A Transformer that transforms a vector to a new feature, which is a sub-array of
    the original feature.It is useful for extracting features from a given vector.

    Note that duplicate features are not allowed, so there can be no overlap between
    selected indices. If the max value of the indices is greater than the size of
    the input vector, it throws an IllegalArgumentException.
    """

    def __init__(self, java_model=None):
        super(VectorSlicer, self).__init__(java_model)

    @classmethod
    def _java_transformer_package_name(cls) -> str:
        return "vectorslicer"

    @classmethod
    def _java_transformer_class_name(cls) -> str:
        return "VectorSlicer"

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/vectorslicer/VectorSlicer.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.ml.feature.vectorslicer;

import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.java.typeutils.RowTypeInfo;
import org.apache.flink.ml.api.Transformer;
import org.apache.flink.ml.common.datastream.TableUtils;
import org.apache.flink.ml.linalg.DenseIntDoubleVector;
import org.apache.flink.ml.linalg.IntDoubleVector;
import org.apache.flink.ml.linalg.SparseIntDoubleVector;
import org.apache.flink.ml.linalg.typeinfo.VectorTypeInfo;
import org.apache.flink.ml.param.Param;
import org.apache.flink.ml.util.ParamUtils;
import org.apache.flink.ml.util.ReadWriteUtils;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
import org.apache.flink.table.api.internal.TableImpl;
import org.apache.flink.types.Row;
import org.apache.flink.util.Preconditions;

import org.apache.commons.lang3.ArrayUtils;

import java.io.IOException;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * A Transformer that transforms a vector to a new feature, which is a sub-array of the original
 * feature. It is useful for extracting features from a given vector.
 *
 * <p>Note that duplicate features are not allowed, so there can be no overlap between selected
 * indices. If the max value of the indices is greater than the size of the input vector, it throws
 * an IllegalArgumentException.
 */
public class VectorSlicer implements Transformer<VectorSlicer>, VectorSlicerParams<VectorSlicer> {
    private final Map<Param<?>, Object> paramMap = new HashMap<>();

    public VectorSlicer() {
        ParamUtils.initializeMapWithDefaultValues(paramMap, this);
    }

    @Override
    public Table[] transform(Table... inputs) {
        Preconditions.checkArgument(inputs.length == 1);
        StreamTableEnvironment tEnv =
                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
        RowTypeInfo inputTypeInfo = TableUtils.getRowTypeInfo(inputs[0].getResolvedSchema());
        RowTypeInfo outputTypeInfo =
                new RowTypeInfo(
                        ArrayUtils.addAll(inputTypeInfo.getFieldTypes(), VectorTypeInfo.INSTANCE),
                        ArrayUtils.addAll(inputTypeInfo.getFieldNames(), getOutputCol()));
        DataStream<Row> output =
                tEnv.toDataStream(inputs[0])
                        .map(new VectorSliceFunction(getIndices(), getInputCol()), outputTypeInfo);
        Table outputTable = tEnv.fromDataStream(output);
        return new Table[] {outputTable};
    }

    @Override
    public void save(String path) throws IOException {
        ReadWriteUtils.saveMetadata(this, path);
    }

    public static VectorSlicer load(StreamTableEnvironment env, String path) throws IOException {
        return ReadWriteUtils.loadStageParam(path);
    }

    @Override
    public Map<Param<?>, Object> getParamMap() {
        return paramMap;
    }

    /**
     * Vector slice function which transforms a vector to a new one with a sub-array of the original
     * features.
     */
    private static class VectorSliceFunction implements MapFunction<Row, Row> {
        private final Integer[] indices;
        private final String inputCol;
        private int maxIndex = -1;

        public VectorSliceFunction(Integer[] indices, String inputCol) {
            this.indices = indices;
            for (Integer index : indices) {
                maxIndex = Math.max(maxIndex, index);
            }
            this.inputCol = inputCol;
        }

        @Override
        public Row map(Row row) throws Exception {
            IntDoubleVector inputVec = row.getFieldAs(inputCol);
            IntDoubleVector outputVec;
            if (maxIndex >= inputVec.size()) {
                throw new IllegalArgumentException(
                        "Index value "
                                + maxIndex
                                + " is greater than vector size:"
                                + inputVec.size());
            }
            if (inputVec instanceof DenseIntDoubleVector) {
                double[] values = new double[indices.length];
                for (int i = 0; i < indices.length; ++i) {
                    values[i] = ((DenseIntDoubleVector) inputVec).values[indices[i]];
                }
                outputVec = new DenseIntDoubleVector(values);
            } else {
                int nnz = 0;
                SparseIntDoubleVector vec = (SparseIntDoubleVector) inputVec;
                int[] outputIndices = new int[indices.length];
                double[] outputValues = new double[indices.length];
                for (int i = 0; i < indices.length; i++) {
                    double val = vec.get(indices[i]);
                    if (val != 0) {
                        outputIndices[nnz] = i;
                        outputValues[nnz] = val;
                        nnz++;
                    }
                }
                if (nnz < outputIndices.length) {
                    outputIndices = Arrays.copyOf(outputIndices, nnz);
                    outputValues = Arrays.copyOf(outputValues, nnz);
                }
                outputVec = new SparseIntDoubleVector(indices.length, outputIndices, outputValues);
            }
            return Row.join(row, Row.of(outputVec));
        }
    }
}
